### PR TITLE
Add Severity and VerboseLevel to grpclog.

### PR DESCRIPTION
**Pull request by menghanl**
updates #120
updates #1044
#1052
#954
#1059

### CODE REVIEW

**Comment by tamird** [comment]
This approach adds an allocation to every logged line 🙍

**Comment by menghanl** [comment]
Reverted the commit to use stringer, including the memory allocation.
The change was intended to add support for structured logging. Will think about that later.

**Comment by menghanl** [comment]
ping

**Comment by dfawley** [comment]
As discussed, please maintain compatibility with older clients by creating a new interface and keeping the old one, and wrapping the old Logger with a new Logger.

File: grpclog/glogger/glogger.go [hunk 1]
```diff
@@ -37,38 +37,64 @@ Package glogger defines glog-based logging for grpc.
...
 import (
-	"fmt"
-
 	"github.com/golang/glog"
 	"google.golang.org/grpc/grpclog"
 )
 
 func init() {
-	grpclog.SetLogger(&glogger{})
+	grpclog.SetLoggerv2(&glogger{})
```
Comment (dfawley) [inline grpclog/glogger/glogger.go]: Nit: LoggerV2, not Loggerv2

File: grpclog/glogger/glogger.go [hunk 2]
```diff
@@ -37,38 +37,64 @@ Package glogger defines glog-based logging for grpc.
 package glogger
```
Comment (dfawley) [inline grpclog/glogger/glogger.go]: How about a usage comment for this package?  "Import this package anonymously to activate it as the enabled grpc logger" or something?

File: grpclog/loggerv2.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
+/*
+ *
+ * Copyright 2015, Google Inc.
```
Comment (dfawley) [inline grpclog/loggerv2.go]: Change the date?

File: grpclog/loggerv2.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	"log"
+	"os"
+)
+
+// VerboseLevel identifies the verbose level used in grpclog.V() function.
+type VerboseLevel int32
+
+// Loggerv2 does underlying logging work for grpclog.
+type Loggerv2 interface {
+	// Info logs args. Arguments are handled in the manner of fmt.Print.
```
Comment (dfawley) [inline grpclog/loggerv2.go]: Info logs args _informationally_?  Or something?

File: grpclog/loggerv2.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+type loggerT struct {
+	m []*log.Logger
+}
+
+// NewLoggerv2 creates a loggerv2 with the provided writers.
+// Fatal logs will be written to errorW, warningW, infoW, followed by exit(1).
+// Error logs will be written to errorW, warningW and infoW.
+// Warning logs will be written to warningW and infoW.
+// Info logs will be written to infoW.
+func NewLoggerv2(infoW io.Writer, warningW io.Writer, errorW io.Writer) Loggerv2 {
```
Comment (dfawley) [inline grpclog/loggerv2.go]: infoW, warningW, errorW io.Writer

File: grpclog/loggerv2_test.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
+/*
+ *
+ * Copyright 2016, Google Inc.
```
Comment (dfawley) [inline grpclog/loggerv2_test.go]: Nit: date.

File: grpclog/grpclog.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	logger.Errorln(args...)
+}
+
+// Fatal is equivalent to Info() followed by a call to os.Exit() with a non-zero exit code.
+func Fatal(args ...interface{}) {
+	logger.Fatal(args...)
+	os.Exit(1)
+}
+
+// Fatalf is equivalent to Infof() followed by a call to os.Exit() with a non-zero exit code.
```
Comment (dfawley) [inline grpclog/grpclog.go]: Please fix this comment to refer to FATAL log.

File: grpclog/loggerv2.go [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+import (
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+)
+
+// VerboseLevel identifies the verbose level used in grpclog.V() function.
+type VerboseLevel int32
```
Comment (dfawley) [inline grpclog/loggerv2.go]: Remove this type and use int directly in V.

File: grpclog/grpclog.go (hunk 7 above)
Comment (menghanl) [inline grpclog/grpclog.go]: done

File: grpclog/loggerv2.go (hunk 8 above)
Comment (menghanl) [inline grpclog/loggerv2.go]: done

**Comment by peter-edge** [comment]
Can we make this more "standard" with regards to popular logging packages, ie github.com/Sirupsen/logrus and github.com/uber-go-zap? Things in particular:

- Rename `Warning` to `Warn`
- Remove the `V` function

I don't think https://github.com/golang/glog was ever meant to be this widely used, and it would be nicer for this to be more in line with existing loggers.

Also to note: do we have to support both `Foo` and `Fooln`? They generally end up doing the same, but this is an arguable point.

**Comment by menghanl** [comment]
@peter-edge 
`github.com/Sirupsen/logrus ` actually has both `Warn` and `Warning`.

And also, we want to use `V` to warp some logs that are off by default.

**Pull request merged**